src/transformers/models/dbrx/modeling_dbrx.py

```diff
@@ -0,0 +1,2132 @@
+    routing_weights = torch.nn.functional.softmax(concatenated_gate_logits,
+                                                  dim=-1)
+
+    _, selected_experts = torch.topk(routing_weights, top_k, dim=-1)
+
+    expert_mask = torch.nn.functional.one_hot(selected_experts, num_experts)
+
+    if attention_mask is None:
+        # Compute the percentage of tokens routed to each experts
+        tokens_per_expert = torch.mean(expert_mask.float(), dim=0)
+
+        # Compute the average probability of routing to these experts
+        router_prob_per_expert = torch.mean(routing_weights, dim=0)
+    else:
+        batch_size, sequence_length = attention_mask.shape
+        num_hidden_layers = concatenated_gate_logits.shape[0] // (
+            batch_size * sequence_length)
+
+        # Compute the mask that masks all padding tokens as 0 with the same shape of expert_mask
+        expert_attention_mask = (attention_mask[None, :, :, None, None].expand(
+            (num_hidden_layers, batch_size, sequence_length, top_k,
+             num_experts)).reshape(-1, top_k, num_experts).to(compute_device))
+
+        # Compute the percentage of tokens routed to each experts
+        tokens_per_expert = torch.sum(
+            expert_mask.float() * expert_attention_mask, dim=0) / torch.sum(
+                expert_attention_mask, dim=0)
+
+        # Compute the mask that masks all padding tokens as 0 with the same shape of tokens_per_expert
+        router_per_expert_attention_mask = (
+            attention_mask[None, :, :, None].expand(
+                (num_hidden_layers, batch_size, sequence_length,
+                 num_experts)).reshape(-1, num_experts).to(compute_device))
+
+        # Compute the average probability of routing to these experts
+        router_prob_per_expert = torch.sum(
+            routing_weights * router_per_expert_attention_mask,
+            dim=0) / torch.sum(router_per_expert_attention_mask, dim=0)
+
+    overall_loss = torch.sum(tokens_per_expert *
+                             router_prob_per_expert.unsqueeze(0))
+    return overall_loss * num_experts
+
+
+#############################################################################
+
+
+def resolve_ffn_act_fn(
+        ffn_act_fn: dict) -> Callable[[torch.Tensor], torch.Tensor]:
+    """Resolve the activation function for the feed-forward network.
+
+    Args:
+        ffn_act_fn (dict): The configuration dictionary for the activation function.
+            The dict config must specify the 'name' of a torch.nn.functional activation
+            function. All of other key values pairs are bound to the function as a partial.
+
+    Returns:
+        Callable[[torch.Tensor], torch.Tensor]: The activation function.
+    """
+    config = deepcopy(ffn_act_fn)
+    name = config.pop('name')
+    if not hasattr(nn.functional, name):
+        raise ValueError(f'Unrecognised activation function name ({name}).')
+    act = getattr(nn.functional, name)
+    return partial(act, **config)
```

Comment:
we should us `from ...activations import ACT2FN` and `ACT2FN`. Feel free to add missing activations in there